src/main/kotlin/com/atlassian/performance/tools/jvmtasks/api/TaskTimer.kt

```diff
@@ -17,12 +23,60 @@ object TaskTimer {
         task: () -> T
     ): T {
         CloseableThreadContext.push(label).use {
+            val path = ThreadContext.cloneStack().asList()
+            taskStartedHandlers.forEach { it.onTaskStarted(path) }
             val start = now()
-            val result = task()
+            val result =
+                try {
+                    task()
+                } catch (e: Exception) {
+                    val duration = Duration.between(start, now())
+                    taskFailedHandlers.forEach { it.onTaskFailed(path, duration, e) }
+                    throw e
+                }
             val duration = Duration.between(start, now())
-
+            taskSucceededHandlers.forEach { it.onTaskSucceeded(path, duration) }
             logger.debug("The task took $duration")
             return result
         }
     }
+
+    fun registerTaskStartedHandler(handler: TaskStartedHandler) {
+        taskStartedHandlers.add(handler)
+    }
+
+    fun registerTaskSucceededHandler(handler: TaskSucceededHandler) {
+        taskSucceededHandlers.add(handler)
+    }
+
+    fun registerTaskFailedHandler(handler: TaskFailedHandler) {
+        taskFailedHandlers.add(handler)
+    }
+
+    fun deregisterTaskStartedHandler(handler: TaskStartedHandler) {
+        taskStartedHandlers.remove(handler)
+    }
+
+    fun deregisterTaskSucceededHandler(handler: TaskSucceededHandler) {
+        taskSucceededHandlers.remove(handler)
+    }
+
+    fun deregisterTaskFailedHandler(handler: TaskFailedHandler) {
+        taskFailedHandlers.remove(handler)
+    }
+
+    @ThreadSafe
+    interface TaskStartedHandler {
+        fun onTaskStarted(path: List<String>)
+    }
+
+    @ThreadSafe
+    interface TaskSucceededHandler {
+        fun onTaskSucceeded(path: List<String>, duration: Duration)
+    }
+
+    @ThreadSafe
+    interface TaskFailedHandler {
```

Comment:
Functional interfaces make it easier to convert from lambdas:
```suggestion
    fun interface TaskStartedHandler {
        fun onTaskStarted(path: List<String>)
    }

    @ThreadSafe
    fun interface TaskSucceededHandler {
        fun onTaskSucceeded(path: List<String>, duration: Duration)
    }

    @ThreadSafe
    fun interface TaskFailedHandler {
```

Reply:
It requires Kotlin 1.4

Reply:
Ah, TIL, thanks!